### ISSUE TITLE
Contrib Group Application: Entity Reference Count

### ISSUE THREAD

**Issue by NormPlum** [backdrop-ops/contrib#845]
Hello and welcome to the contrib application process! We're happy to have you :)

**Please indicate how you intend to help the Backdrop community by joining this group**
* Option 1: I would like to contribute a project

## Based on your selection above, please provide the following information:

**(option 1) The name of your module, theme, or layout**

Entity Reference Count

## (option 1) Please note these 3 requirements for new contrib projects:

- [x] Include a README.md file containing license and maintainer information.
      You can use this example: https://raw.githubusercontent.com/backdrop-ops/contrib/master/examples/README.md
- [x] Include a LICENSE.txt file.
      You can use this example: https://raw.githubusercontent.com/backdrop-ops/contrib/master/examples/LICENSE.txt.
- [x] If porting a Drupal 7 project, Maintain the Git history from Drupal.

**(option 1 -- optional) Post a link here to an issue in the drupal.org queue notifying the Drupal 7 maintainers that you are working on a Backdrop port of their project**

No issue, as I don't have a Drupal account. But here's the project: https://www.drupal.org/project/entityreference_count

**Post a link to your new Backdrop project under your own GitHub account (option 1)**

https://github.com/NormPlum/entityreference_count

**If you have chosen option 2 or 1 above, do you agree to the [Backdrop Contributed Project Agreement](https://github.com/backdrop-ops/contrib#backdrop-contributed-project-agreement)**

YES

**Comment by herbdool** [comment]
@NormPlum this all looks good to me: you've agreed to the argreement; your module checks all the boxes. So I think you're good to go.

@klonos @jenlampton

**Comment by jenlampton** [comment]
I agree with @herbdool - since this application meets all 3 requirements, **an invitation to join the Backdrop Contrib group is on the way!**

Once you have accepted the invitation, feel free to [transfer the repository](https://docs.backdropcms.org/documentation/after-your-application-is-accepted) into the backdrop-contrib group at any time (ask here if you have questions).

Welcome to contrib, we are happy to have you!

---

I'm going to do a quick code review in the next comment - Take or leave the suggestions as you see fit! We do this review specifically looking for things that may be done differently between Backdrop and Drupal. We hope that this review might help with the transition for those new to Backdrop :)

**Comment by jenlampton** [comment]
First: this module looks great - I don't know how much you changed from the Drupal version, but all the function definitions + docblocks are immaculate, so, thank you :)  

I don't see anything Drupal vs Backdrop specific that needs to be changed, but I have a few notes below about possible improvements User experience, and Performance - two of Backdrop's guiding principles. 

Install file:
- Git prefers when we leave a blank line at the very end of the file. If you added one to this file it would make Git complain less for anyone who ends up contributing :)

Module file:
- In `entityreference_count_field_instance_settings_form()` there is a variable named `$er_fields`.  We usually don't abbreviate anything in variable names so that it is easier for other developers to understand what they are for (but in this case `$entityreference_fields` is very long - which is probably why it was shortened in Drupal). I probably would have named this $reference_fields 🤷 
- The form element `counted_reference_fields` has both a title (`Entity reference fields that may be counted:`) and a description (`Select the entity reference fields that you would like to count.`), but the description doesn't add any value. This is very common in Drupal, but in Backdrop we have a focus on user experience, so I usually "reduce the UI clutter" by deleting these repetitive descriptions. 
- For the elements `singular` and `plural` the description starts by repeating the label, I would remove `The string for the singular case.` and `The string for the plural case.` since that much is clear from the labels. The UI is also much friendlier if we can avoid using words like "string" when it's not critical :)  
- For the title on `link` (`Link count to the referenced entity`) I would remove the word "count". From the current sentence alone it's not clear that "Link" is being used as a verb and not part of the noun "Link count".  Additionally, the text `Link to the referenced entity` is used below in `entityreference_count_field_formatter_settings_summary()` - so making them match will reduce the number of strings that need to be translated :)
- There is a line commented out on 533, it looks like that can probably be removed?
- I would move the function `entityreference_count_entityreference_count_updated()` into entityreference_count.rules.inc since that file will only be included if the module exists. 
- I'm not sure why `hook_exit()` is being used to count the references (I'm sure there's a use case, or maybe a bug that arose in the Drupal module?). It seems like counting is something that should be done only when the count is likely to change, and not on every page view. I'd leave this alone for now, but doing database writes on page view is a performance issue, so I'd probably create an issue to see if there's an alternative way to solve this.
- Again with the blank line at the end of the file :) 

Rules Include:
- See note above about moving any rules-specific hook implementations here - this would be a micro-performance enhancement.
- Again with the blank line at the end of the file :) 

Thank you for taking the time to contribute this project to Backdrop!

**Comment by jenlampton** [comment]
Thanks @NormPlum ! I saw the module moved to contrib 

> I'm a bit confused about the blank line at end of file suggestion

If Git is not complaining, then this is a non-issue. Thank you for looking into it and I'm sorry if I wasted your time. 

> Most of the other suggestions were things that Drupal did,

I figured, and it's perfectly fine to leave them all as-is. Just writing down my thoughts as I looked over the code :) 

> I researched where hook implementations need to live for my own code, and I think this doesn't work. I tested by calling module_implements('entityreference_count_updated') before and after moving the function to the Rules file. Before, it worked. After, it didn't.

Could be!  Just to sanity check - you did have rules module enabled in the "after" scenario? If so, I wonder how rules loads that .inc file. It's also possible my memory is out-of date! 

> This was from Drupal and I'm not sure I know a better way to do it.

I'm not sure I do either, but it seems like there should be a post-update if we can't do a query in update. Hook exit is never the "right" answer -- but this could be an API limitation we should examine for core.
